File: Data Structures/4 Queue/3  Deque Library/Deque Library.py
from collections import deque

"""
    Demonstrates the usage of `collections.deque` as a queue in Python.

    - While `deque` can be used for strict FIFO (First In, First Out) behavior, it also allows for flexible double-ended operations.
    - `deque` supports **both** FIFO and reverse (LIFO: Last In, First Out) operations efficiently, with O(1) time complexity for operations at both ends, unlike the `queue.Queue` library.

    Basic Operations:
    - **Append**: `append(item)` - Adds an element to the rear of the queue (FIFO).
    - **Popleft**: `popleft()` - Removes and returns the front element of the queue (FIFO).
    - **Appendleft**: `appendleft(item)` - Adds an element to the front of the queue (used for reverse or LIFO operations).
    - **Pop**: `pop()` - Removes and returns the last element of the queue (LIFO).
    - **Peek**: `queue[0]` - Peeks at the front element of the queue (FIFO).
    - **Back Peek**: `queue[-1]` - Peeks at the back element of the queue (LIFO).
"""


# Create a queue using deque
queue = deque()

# ----- Strict FIFO Queue Operations -----

queue.append("a")   # Enqueue (add to rear)
queue.append("b")
queue.append("42")

print("Queue after enqueues:", list(queue))  # ['a', 'b', '42']

print("Front Peek:", queue[0])  # Peek front (equivalent to front())

queue.popleft()  # Dequeue (remove from front)
print("Queue after dequeue:", list(queue))  # ['b', '42']

# ----- Reverse Operations -----

queue.appendleft("X")  # Enqueue at front
print("Queue after enqueue_front:", list(queue))  # ['X', 'b', '42']

print("Back Peek:", queue[-1])  # Peek back (equivalent to back())

queue.pop()  # Dequeue from back
print("Queue after dequeue_back:", list(queue))  # ['X', 'b']

File: Data Structures/4 Queue/3  Deque Library/Deque Library.cpp
#include <iostream>
#include <deque>

using namespace std;

/*
    Demonstrates the usage of `std::deque` as a queue in C++.

    - While `std::deque` supports strict FIFO (First In, First Out) behavior, it also allows for flexible double-ended operations.
    - `std::deque` supports **both** FIFO and reverse (LIFO: Last In, First Out) operations efficiently, with O(1) time complexity for insertions and deletions at both ends.
    - Unlike `std::queue`, which only allows push at the back and pop at the front, `std::deque` provides more flexibility.

    Basic Operations:
    - **push_back(item)** → Adds an element to the rear of the queue (FIFO).
    - **pop_front()** → Removes the front element (FIFO).
    - **push_front(item)** → Adds an element to the front of the queue (used for reverse or LIFO operations).
    - **pop_back()** → Removes the last element (LIFO).
    - **front()** → Retrieves the front element without removing it (FIFO peek).
    - **back()** → Retrieves the last element without removing it (LIFO peek).
    - **size()** → Returns the number of elements in the deque.
*/

void printDeque(const deque<string>& dq) {
    if (dq.empty()) {
        cout << "Deque is empty!" << endl;
        return;
    }
    for (const auto& item : dq) {
        cout << item << " ";
    }
    cout << endl;
}

int main() {
    deque<string> dq;

    // ----- Strict FIFO Queue Operations -----
    dq.push_back("a");  // Enqueue (add to rear)
    dq.push_back("b");
    dq.push_back("42");

    cout << "Queue after enqueues: ";
    printDeque(dq);  // Output: a b 42

    cout << "Front Peek: " << dq.front() << endl;  // Peek front (equivalent to front())

    dq.pop_front();  // Dequeue (remove from front)
    cout << "Queue after dequeue: ";
    printDeque(dq);  // Output: b 42

    // ----- Reverse Operations -----
    dq.push_front("X");  // Enqueue at front
    cout << "Queue after enqueue_front: ";
    printDeque(dq);  // Output: X b 42

    cout << "Back Peek: " << dq.back() << endl;  // Peek back (equivalent to back())

    dq.pop_back();  // Dequeue from back
    cout << "Queue after dequeue_back: ";
    printDeque(dq);  // Output: X b

    return 0;
}
